### ISSUE TITLE
Camera on iOS - stopRecording intermittently doesn’t cause recordAsync to complete

### ISSUE THREAD

**Issue by DanBruns** [expo/expo#2837]
Environment:
  OS: Windows 10
  Node: 8.11.2
  Yarn: Not Found
  npm: 5.6.0
  Watchman: Not Found
  Xcode: N/A
  Android Studio: Version  3.2.0.0 AI-181.5540.7.32.5014246

Packages: (wanted => installed)
  expo: ^31.0.4 => 31.0.5
  react: 16.5.0 => 16.5.0
  react-native: https://github.com/expo/react-native/archive/sdk-31.0.1.tar.gz => 0.57.1

Diagnostics report:
  https://exp-xde-diagnostics.s3.amazonaws.com/dbruns-f55e9ee7-6682-41da-90c7-f4b4e3522bb3.tar.gz

We are targeting both iOS and Android, 

We’re having intermittent issues with recording video on iOS, but not with all devices. It seems to work on the latest generation iPhones (iPhone X running iOS 12.1), but not on older iPad (10.3.3) or iPhones (iPhone 6 running iOS 12.1)  When it fails, the behavior is and can be reproduced as follows:

1. start recording via recordAsync

2. call stopRecording. Nothing happens.

3. call stopRecording a second time. The promise resolves and the video is available, but it includes only the segment between the two calls to stopRecording.

Sometimes, on the first call to stopRecording, the camera preview flashes, making us think this is when the recording is actually starting.

Here’s a snack of a simple video record/playback app: 
https://snack.expo.io/@dbruns/ios-camera-recordasync-issue

When we go through the process of recording, then pressing “reset” (which clears the completed state info), and recording again, repeating several times, often the process fails every other time and succeeds the other times.

Expected Behavior

We expect that camera would start recording when recordAsync is called (or very shortly thereafter) and will stop, with the recordAsync promise resolved, when stopRecording is called, providing us with the resulting video.

Actual Behavior

As described above, we often have to call stopRecording twice (on some iOS devices) to get a video, and even then, the video isn't for the time interval expected.

Reproducible Demo

https://snack.expo.io/@dbruns/ios-camera-recordasync-issue

**Comment by ColinJS** [comment]
We've got the same issue but with iOS X also (we've tried on X and 6). 
You can reproduce it by following those steps:
-expo init
-paste this in App.js : 
```
import React from 'react';
import {
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
  Alert
} from 'react-native';
import { Camera, Permissions } from 'expo';

export default class HomeScreen extends React.Component {
  static navigationOptions = {
    header: null,
  };

  constructor(props){
    super(props)
    this.state = {
      hasVideoPermission: false,
      hasAudioPermission: false,
    }
    this.recordingTest = false
  }

  componentDidMount(){
    this.askPermissions()
  }

  askPermissions = async () => {
    const { status: status_video } = await Permissions.askAsync(Permissions.CAMERA);
    const { status: status_audio } = await Permissions.askAsync(Permissions.AUDIO_RECORDING);
    this.setState({ hasVideoPermission: status_video === 'granted', hasAudioPermission: status_audio === 'granted' });
  }

  toggleCameraRecording = () => {
    if(this.recordingTest){
      this._camera.stopRecording()
    }else{
      Alert.alert('start recording')
      this.recordingTest = true
      this._camera.recordAsync().then(res => {
        Alert.alert('stop recording')
        this.recordingTest = false
      })
    }
  }

  render() {
    return (
      <View style={styles.container}>
        { this.state.hasAudioPermission && this.state.hasVideoPermission && <Camera
          ref={ref=>{
            this._camera = ref
          }}
          style={styles.camera}
          autoFocus={true}
          flashMode={'off'}
          type={'back'}
          ratio={'16:9'}
        >
        <TouchableOpacity style={{height: '100%', width: '100%', justifyContent: 'center', alignItems: 'center'}} onPress={this.toggleCameraRecording}>
          <Text>Click</Text>
        </TouchableOpacity>
        </Camera>}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
  },
  camera: {
    flex: 1,
    justifyContent: 'space-between',
  },
});
```
-expo start

And then try it with any iOS device. 

On my project I find out that it's not only by pressing on the stopRecord button a second time that it fired the stopRecording. If you press on any button that do something after the first stopRecording it does fire the stopRecording.
So it's a bit like the stopRecording is waiting for an action to fire. I tried to force a rerender but it does nothing, putting a setTimeout does nothing neiter and using stopRecording two time does nothing.

**Comment by ammarBlueberry** [comment]
Hi, I have essentially the same as issue as you guys. Except in my case, this behavior only manifests itself in the iOS app binary. 
In my case, I have to press the button twice to stop recording the video and the footage is from when I pressed the button to start recording to when when I tapped the button the second time. 

I guess another way of looking at this is that the recordAsync promise does not resolve when stopRecording is called first.
